completions/xrandr

```diff
@@ -1,62 +1,165 @@
+            fi
             return
             ;;
         --rotate)
-            COMPREPLY=($(compgen -W 'normal inverted left right' -- "$cur"))
+            if [[ -v output ]]; then
+                COMPREPLY=($(compgen -W 'normal inverted left right' -- "$cur"))
+            fi
+            return
+            ;;
+        --filter)
+            if [[ -v output ]]; then
+                COMPREPLY=($(compgen -W 'bilinear nearest' -- "$cur"))
+            fi
             return
             ;;
-        --setprovideroutputsource | --setprovideroffloadsink)
-            local providers=$("$1" --listproviders 2>/dev/null |
-                command sed -ne 's/.* name:\([^ ]*\).*/\1/p')
+        --setprovideroutputsource)
+            local providers=$(_xrandr_providers "$1" "Sink Output")
             COMPREPLY=($(compgen -W "$providers" -- "$cur"))
-            # TODO 2nd arg needed, is that a provider as well?
+            return
+            ;;
+        --setprovideroffloadsink)
+            local providers=$(_xrandr_providers "$1" "Source Offload")
+            COMPREPLY=($(compgen -W "$providers" -- "$cur"))
+            return
+            ;;
+        --delmonitor)
+            local monitors=$(_xrandr_monitors "$1")
+            COMPREPLY=($(compgen -W "$monitors" -- "$cur"))
             return
             ;;
     esac
 
-    COMPREPLY=($(compgen -W '$("$1" -help 2>&1 |
-        command sed -e "s/ or / /g" -e "s/<[^>]*>]//g" | _parse_help -)' \
-        -- "$cur"))
+    # second arguments
+    if [[ $cword -ge 2 ]]; then
+       case "${words[cword-2]}" in
+           --set)
+               return
+               ;;
+           --addmode | --delmode)
+               local modes=$(_xrandr_modes "$1" "${words[cword-1]}")
+               COMPREPLY=($(compgen -W "$modes" -- "$cur"))
+               return
+               ;;
+           --setmonitor)
+               COMPREPLY=($(compgen -W 'auto' -- "$cur"))
+               return
+               ;;
+           --setprovideroutputsource)
+               local providers=$(_xrandr_providers "$1" "Source Output")
+               COMPREPLY=($(compgen -W "0x0 $providers" -- "$cur"))
+               return
+               ;;
+           --setprovideroffloadsink)
+               local providers=$(_xrandr_providers "$1" "Sink Offload")
+               COMPREPLY=($(compgen -W "0x0 $providers" -- "$cur"))
+               return
+               ;;
+       esac
+    fi
+
+    # third arguments
+    if [[ $cword -ge 3 ]]; then
+        case "${words[cword-3]}" in
+            --setmonitor)
+                local outputs=$(_xrandr_outputs "$1")
+                COMPREPLY=($(compgen -W "none $outputs" -- "$cur"))
+                # TODO: the third argument is actually a comma-separated list
+                return
+                ;;
+        esac
+    fi
+
+    local options
+    if [[ -v output ]]; then
+        options=`_xrandr_help "$1" | sed -e 's/ or /\n  /g' -e 's/<[^>]*>]//g' | _parse_help -`
```

Comment:
`$(...)` instead of backticks here and below

Reply:
Done.